flink-end-to-end-tests/flink-sql-client-test-kafka-0.10/pom.xml

```diff
@@ -0,0 +1,165 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```

Comment:
We don't need a Kafka-version specific `pom.xml`. The `flink-sql-client-test/pom.xml` is sufficient. It shouldn't be a problem with both Kafka versions there. The module makes sure that all transitively required Flink modules are build and copies the SQL jars over into a common directory using Maven.

The dependency clashes only occur when executing the SQL Client test. This means that the test should exclude clashing jars for SQL statement execution. This has nothing to do with Maven but is a pure test script thing.

Reply:
@twalthr I think distinguishing the old different versions of the kafka connector will make it clearer (just like testing the kafka connector end-to-end). It is difficult to distinguish dependencies easily in a pom before I split it. Currently we use `<classifier></classifier>` to classify, but it seems that we can't add two `<classifier />` elements to a dependency.

Reply:
Distinguishing the different versions in the test scripts makes sense but we don't need to do it in the Maven module as this Maven module is just responsible for collecting jars. Different Kafka versions were in this `pom.xml` before without any problems.

What do you mean with two `classifier` elements?

Reply:
@twalthr This [plugin](https://github.com/apache/flink/blob/master/flink-end-to-end-tests/flink-sql-client-test/pom.xml#L134) collect jar based on the `classifier `. 

We should collect different sets of dependencies for different kafka connectors. We can't test different kafka connectors but rely on multiple kafka connectors, which can cause multiple versions of dependency conflicts and you can't guarantee that you are using a 0.10 kafka client when testing kafka connector 0.10. There is no guarantee that the Kafka client 2.0 will be tested when testing kafka connector 2.0. In fact, it may have a combination of kafka connector 2.0 and kafka client 0.10. cc @pnowojski

Reply:
Yes, the plugin that you mentioned is responsible for collecting the jars. However, the dependencies in the `pom.xml` just force Maven to transitively compile the fat SQL jars in their modules. We need to add the dependencies for cases when people just build the `flink-end-to-end` module such that all transitive modules are built as well.

Dependency clashes can only occur in the SQL Client test script, but this can be easily handled by explicitly specify which jars to use with `--jar`. If this were a regular Maven project you would be right, but this is more of a helper Maven project, not a typical Flink job project.

Reply:
@twalthr Thanks for your suggestion, I have modified its implementation. What do you think of the current implementation? cc @pnowojski

Reply:
For me, basing on the description from @twalthr, it looks good. I would adhere here to @twalthr judgement. 

The important bit is that end-to-end tests should reflect as closely as possible the exactly same steps as would user perform when running their jobs/queries. Thus in case of data streams end-to-end tests, splitting the example jobs into separate modules with different poms made sense. Here as @twalthr said, this module is only a helper module, while the actual important thing that we want to test is to pass correct "fat jars" to the sql client.